oak_functions/examples/abitest/module/src/lib.rs

```diff
@@ -0,0 +1,119 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Oak Functions ABI test.
+
+use anyhow::{anyhow, Context};
+use assert_matches::assert_matches;
+use oak_functions_abitest_common::*;
+use std::collections::HashMap;
+
+type TestFn = fn(&str) -> anyhow::Result<()>;
+
+struct TestManager {
+    tests: HashMap<String, TestFn>,
+}
+
+impl TestManager {
+    fn new() -> Self {
+        let mut tests = HashMap::new();
+        tests.insert(TEST_READ_WRITE.to_string(), Self::test_read_write as TestFn);
+        tests.insert(
+            TEST_DOUBLE_READ.to_string(),
+            Self::test_double_read as TestFn,
+        );
+        tests.insert(
+            TEST_DOUBLE_WRITE.to_string(),
+            Self::test_double_write as TestFn,
+        );
+        tests.insert(TEST_WRITE_LOG.to_string(), Self::test_write_log as TestFn);
+        tests.insert(
+            TEST_STORAGE_GET.to_string(),
+            Self::test_storage_get as TestFn,
+        );
+
+        Self { tests }
+    }
+
+    fn run_test(&self) -> anyhow::Result<()> {
+        let request = oak_functions::read_request()
+            .map_err(|error| anyhow!("Couldn't read request: {:?}", error))?;
+        let test_name = String::from_utf8(request).context("Couldn't parse request")?;
+        let test = self
+            .tests
+            .get(&test_name)
+            .ok_or(anyhow!("Couldn't find test: {}", &test_name))?;
+        test(&test_name).context(format!("Couldn't run test: {}", &test_name))
+    }
+
+    fn test_read_write(_request: &str) -> anyhow::Result<()> {
+        let result = oak_functions::write_response(TEST_READ_WRITE_RESPONSE.as_bytes());
+        assert_matches!(result, Ok(_));
```

Comment:
Out of curiosity: What happens if any of the asserts in these tests fail? Have you checked that?

Reply:
Looks like, these do not produce a failure on the server side
Cite: "if there is an externally visible effect of a panic inside Wasm it is a potential route for leaking information"
Leaving it as is, will be fixed as part of the https://github.com/project-oak/oak/issues/1963